### PR TITLE
Save the down synced resources of a page as it is dowloaded.

### LINKED ISSUE #862: Device goes OOM when downloading a lot of resources.
**Describe the bug**
Syncing a large number of resources causes OOM issue in the engine. 

**Component**
Core library, SDC library, reference app, or SDC gallery app

**To Reproduce**
1. Setup the server with a large set of data (some resources should atleast 50,000 in number)
2. Sync the resource data to the client.

**Expected behavior**
Data should sync and get stored in the database.

**Actual behavior**
 App crashes because of O.O.M.

**Screenshots**
[OOM_crash_log.txt](https://github.com/google/android-fhir/files/7409748/OOM_crash_log.txt)

**Smartphone (please complete the following information):**
 - Device: [e.g. Pixel4a emulator]
 - Android version: [e.g. Settings -> About phone -> Android version]
 - Build number: [e.g. Settings -> About phone -> Build number]

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

**Side note**
Instead of downloading all the resources and then inserting them to the database, maybe be we should insert downloaded page before we fetch a new one.

### PR DESCRIPTION
**Pull request by aditya-07**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #862

**Description**
While  syncing a large number of resources, the device goes OOM because engine keeps adding the downloaded list to a  master list for saving all the resources in the database later after all resources have been downloaded. 
To mitigate this issue, the changes here save each page as it is downloaded so that the net memory footprint of the app doesn't increase.

**Alternative(s) considered**
NO

**Type**
Choose one: Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/FhirEngine.kt [hunk 1]
```diff
@@ -70,9 +70,13 @@ interface FhirEngine {
 
   /**
    * Synchronizes the [download] result in the database. The database will be updated to reflect the
-   * result of the [download] operation.
+   * result of the [download] operation. [onPageDownloaded] is called with the resources after each
+   * successful download of page.
    */
-  suspend fun syncDownload(download: suspend (SyncDownloadContext) -> List<Resource>)
+  suspend fun syncDownload(
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/FhirEngine.kt]: The api is updated now. Please take a look.

File: engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt [hunk 2]
```diff
@@ -129,9 +130,7 @@ class FhirEngineImplTest {
 
   @Test
   fun syncDownload_downloadResources() = runBlocking {
-    fhirEngine.syncDownload {
-      return@syncDownload listOf(TEST_PATIENT_2)
-    }
+    fhirEngine.syncDownload { flow { emit(listOf(TEST_PATIENT_2)) } }
```
Comment (yigit) [inline engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:133]:
```suggestion
    fhirEngine.syncDownload { flowOf(listOf(TEST_PATIENT_2)) }
```

**Review by yigit — APPROVED** [review]
lgtm. just a small suggestion.

File: engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt (hunk 2 above)
Comment (aditya-07) [reply engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt]: Done.

**Review by jingtang10 — APPROVED** [review]

**Pull request merged**